libs/header/p2p/session.go

```diff
@@ -139,30 +145,50 @@ func (s *session[H]) doRequest(
 
 	r, size, duration, err := sendMessage(ctx, s.host, stat.peerID, s.protocolID, req)
 	if err != nil {
-		log.Errorw("requesting headers from peer failed", "failed peer", stat.peerID, "err", err)
-		select {
-		case <-s.ctx.Done():
-		case s.reqCh <- req:
-			stat.decreaseScore()
-			log.Debug("retrying the request from different peer")
-		}
-		return
+		// we should not punish peer at this point and should try to parse responses, despite that error was received.
+		log.Debugw("requesting headers from peer failed", "failed peer", stat.peerID, "err", err)
 	}
 
 	h, err := s.processResponse(r)
 	if err != nil {
 		switch err {
 		case header.ErrNotFound:
+			// errNoResponses means that we did not fetch any headers from the peer, so we can decrease its score.
+		case errEmptyResponse:
+			stat.decreaseScore()
 		default:
 			s.peerTracker.blockPeer(stat.peerID, err)
+			log.Errorw("processing headers response failed", "failed peer", stat.peerID, "err", err)
 		}
 		select {
 		case <-s.ctx.Done():
 		case s.reqCh <- req:
 		}
 		return
 	}
-	log.Debugw("request headers from peer succeeded", "peer", stat.peerID, "amount", req.Amount)
+
+	log.Debugw("request headers from peer succeeded",
+		"peer", stat.peerID,
+		"receivedAmount", len(h),
+		"requestedAmount", req.Amount,
+	)
+
+	// ensure that we received the correct amount of headers.
+	if uint64(len(h)) != req.Amount {
+		from := uint64(h[len(h)-1].Height())
+		amount := req.Amount - from
+		select {
+		case <-s.ctx.Done():
+			// we should update stats anyway in this case
+			stat.updateStats(size, duration)
+			return
+		// create a new request with the remaining headers.
+		// prepareRequests will return a slice with 1 element at this point
+		case s.reqCh <- prepareRequests(from+1, amount, req.Amount)[0]:
```

Comment:
IIRC, we discussed splitting the requeest up into more smaller pieces or did we cancel that?

Reply:
We did. But then we invented this case where  the amount will be dynamically change during requests, so now it's not necessary to split it into smaller pieces